### ISSUE TITLE
Indexed access types are too eager in evaluation

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#14944]
I have a function that takes a string (`propName`) and returns another function that takes an object (`obj`), and fetches the value of the property named by the original string (`obj[propName]`). In other words:

```ts
function createPlucker<K extends string>(propName: K) {

  return function plucker<T extends Record<K, {}>>(obj: T): T[K] {
    let result = obj[propName];
    return result;
  };
}

// A function which retrieves properties named 'x' off of objects.
const pluckX = createPlucker('x');

const point = {
  x: 100,
  y: 100,
};

// Should be a 'number'
const test = pluckX(point);

test.toFixed();
```

**Expected**: No error, `test` should have the type `number`.
**Actual**: `toFixed` does not exist on `{}`. `test has the type `{}`.

**Comment by DanielRosenwasser** [comment]
Context: https://twitter.com/appsforartists/status/847691492894191619

**Comment by appsforartists** [comment]
Incidentally, here's the actual code for `createPlucker`:
```javascript

**Comment by ahejlsberg** [comment]
@DanielRosenwasser Your example works as expected with the compiler in master. Which version are you using?

**Comment by appsforartists** [comment]
Can confirm it works in TypeScript 2.3; however, it breaks if you try to put a return type on `createPlucker`:

```typescript
export function createPlucker<K extends string, T extends Record<K, {}>>(path: K): (value:T) => T[K] {
  const pathSegments = path.split('.');

  return function plucker(value: T): T[K] {
    let result: T[K] = value;

    for (let pathSegment of pathSegments) {
      result = result[pathSegment];
    }

    return result;
  };
}

const pluckX = createPlucker('x');
const point = { x: 0, y: 100 };
const x = pluckX(point);
x.toFixed();
```

I get `Property 'toFixed' does not exist on type '{}'`.  Before trying to type the return value of `createPlucker`, `x` is correctly inferred to be `number`.

(Note, I'd like to be able to abstract Plucker into its own type, shown below.  I've inlined it in the example above to minimize the number of abstractions.)

```typescript
export type Plucker<K extends string, T extends Record<K, {}>> = (value: T) => T[K];
```
